### PR TITLE
feat: New truncate-heading prop for header

### PR DESCRIPTION
**Pull request by pan-kot**
### Description

Default heading wrapping behavior might not be applicable when used inside small containers. When new property is set, the heading is truncated with ellipsis instead.

When the heading is truncated some part of its information is lost (exception - screen-reader users). That means the feature can only be used when there is an alternative way to obtain it.

### How has this been tested?

Screenshot tests

<details>
   <summary>Review checklist</summary>

_The following items are to be evaluated by the author(s) and the reviewer(s)._

#### Correctness

- _Changes include appropriate documentation updates._
- _Changes are backward-compatible if not indicated, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#public-apis)._
- _Changes do not include unsupported browser features, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#browsers-support)._
- _Changes were manually tested for accessibility, see [accessibility guidelines](https://cloudscape.design/foundation/core-principles/accessibility/)._

#### Security

- _If the code handles URLs: all URLs are validated through [the `checkSafeUrl` function](https://github.com/cloudscape-design/components/blob/main/src/internal/utils/check-safe-url.ts)._

#### Testing

- _Changes are covered with new/existing unit tests?_
- _Changes are covered with new/existing integration tests?_
</details>

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/header/interfaces.ts [hunk 1]
```diff
@@ -41,6 +41,11 @@ export interface HeaderProps extends BaseComponentProps {
    * Area next to the heading to display an Info link.
    */
   info?: React.ReactNode;
+  /**
+   * Determines if the heading (specified with the `children` property) should be truncated with ellipsis when there is not enough space.
+   * Defaults to `false`.
+   */
+  truncateHeading?: boolean;
```
Comment (just-boris) [inline src/header/interfaces.ts:48]: Naming. 

1. We have similar props on button component (called `wrapText`) and table (`wrapLines`). So the new property should follow this pattern
2. The content of default slot is called `title` (see display name in the [API docs](https://github.com/cloudscape-design/components/blob/1b69f03545f9f0aaf88fb685e28efabf1ca05140/src/header/interfaces.ts#L10) and also [usage guidelines](https://cloudscape.design/components/header/?tabId=usage)).

Based on these two, I suggest `wrapTitle` for the name

Comment (just-boris) [reply src/header/interfaces.ts:48]: P.S. you may also look at API docs for respective components and reuse some of the wording, but I do not have much opinion there

File: src/header/internal.tsx [hunk 2]
```diff
@@ -60,10 +62,20 @@ export default function InternalHeader({
...
+              truncateHeading && styles['heading-overflow']
+            )}
+          >
+            <span
+              className={clsx(styles['heading-text'], styles[`heading-text-variant-${variantOverride}`])}
+              aria-hidden={truncateHeading}
+            >
               {children}
             </span>
+            {truncateHeading && <ScreenreaderOnly>{children}</ScreenreaderOnly>}
```
Comment (just-boris) [inline src/header/internal.tsx:78]: Why is this needed, what is wrong with the visible text?

`ScreenreaderOnly` component works as maximum truncation down to 1px, so what is wrong with relying on half-truncated content?

File: src/header/styles.scss [hunk 3]
```diff
@@ -103,6 +103,10 @@
 }
 
 .title {
+  display: flex;
```
Comment (just-boris) [inline src/header/styles.scss:106]: Remembering the story with flexible containers (https://github.com/cloudscape-design/components/pull/603) can we make extra flex-styles opt-in for the new mode to reduce possible impact?

File: src/header/internal.tsx (hunk 2 above)
Comment (DaemonCahill) [reply src/header/internal.tsx:78]: I just bumped into this as well when I noticed that there are two DOM nodes with the same text when aria-hidden='true'.

Screen Shot 2023-02-13 at 4 05 34 PM

I don't think I understand what problem is solved by putting aria-hidden on the truncated node and then replicating the same text and rendering it off the canvas...

**Comment by DaemonCahill** [comment]
It might be worthwhile to expand the screenshot testing of this property beyond just rendering headers.

qa

File: src/header/interfaces.ts (hunk 1 above)
Comment (pan-kot) [reply src/header/interfaces.ts:48]: Thanks. 

The reason I used `heading` and not `title` is because title also contains counter and info link that are not truncated.

The reason I used `truncate` and not `wrap` is for default value for that property to be `false`.

I am happy to challenge both decisions - I'll bring that to the team review.

File: src/header/internal.tsx (hunk 2 above)
Comment (pan-kot) [reply src/header/internal.tsx:78]: That allows screen-reader users to always read the full contents no matter if the actual header is truncated or not.

I don't mind removing it for now and adding back in case we'll have the feedback.

File: src/header/styles.scss (hunk 3 above)
Comment (pan-kot) [reply src/header/styles.scss:106]: Yes, I think it is better to do that.

Ideally is to have fewer conditions in the code but I agree it is better to decrease the blast radius. I am still waiting for the screenshot test results to see if there are noticeable diffs.

**Comment by pan-kot** [comment]
> It might be worthwhile to expand the screenshot testing of this property to more diverse use cases than just rendering standalone headers.
> 
> qa

The feature is specifically introduced for dashboard and I agree it does make total sense to test it there as well. As for other possible use cases I'd hesitate. The problem with truncation is that it hides certain information and there must be a way to recover it. Otherwise the feature is not applicable.

File: src/header/internal.tsx (hunk 2 above)
Comment (DaemonCahill) [reply src/header/internal.tsx:78]: Pardon my ignorance but isn't the opposite true? As long as aria hidden is not set, the presence of hidden overflow will still announce the entire content to a screenreader?

**Comment by DaemonCahill** [comment]
> The feature is specifically introduced for dashboard and I agree it does make total sense to test it there as well. As for other possible use cases I'd hesitate.

@pan-kot Once we release this it will be a public API on the header component that can be used anywhere, right? So, the intended use behind the API is somewhat irrelevant and what matters is where it _could_ be used?

**Comment by pan-kot** [comment]
Closed due to a better mitigation found for board components.
